go/pvl/interp.go

```diff
@@ -0,0 +1,911 @@
+				"Selector entry string or int %v", selector)
+		}
+	}
+
+	return selection, nil
+}
+
+// Most failures here log instead of returning an error. If an error occurs, ([], nil) will be returned.
+// This is because a selector may descend into many subtrees and fail in all but one.
+func pvlRunSelectorJSONInner(g ProofContextExt, state PvlScriptState, object *jsonw.Wrapper, selectors []*jsonw.Wrapper) ([]string, libkb.ProofError) {
+	// The terminating condition is when we've consumed all the selectors.
+	if len(selectors) == 0 {
+		s, err := pvlJSONStringSimple(object)
+		if err != nil {
+			debugWithState(g, state, "JSON could not read object: %v (%v)", err, object)
+			return make([]string, 0), nil
+		}
+		return []string{s}, nil
+	}
+
+	selector := selectors[0]
+	nextselectors := selectors[1:]
+
+	selectorIndex, err := selector.GetInt()
+	selectorIsIndex := err == nil
+	selectorKey, err := selector.GetString()
+	selectorIsKey := err == nil && !selectorIsIndex
+	allness, err := selector.AtKey("all").GetBool()
+	selectorIsAll := err == nil && allness
+
+	switch {
+	case selectorIsIndex:
+		object, err := object.ToArray()
+		if err != nil {
+			debugWithState(g, state, "JSON select by index from non-array: %v (%v) (%v)", err, selectorIndex, object)
+			return []string{}, nil
+		}
+
+		nextobject := object.AtIndex(selectorIndex)
+		return pvlRunSelectorJSONInner(g, state, nextobject, nextselectors)
+	case selectorIsKey:
+		object, err := object.ToDictionary()
+		if err != nil {
+			debugWithState(g, state, "JSON select by key from non-map: %v (%v) (%v)", err, selectorKey, object)
+			return []string{}, nil
+		}
+
+		nextobject := object.AtKey(selectorKey)
+		return pvlRunSelectorJSONInner(g, state, nextobject, nextselectors)
+	case selectorIsAll:
+		children, err := pvlJSONGetChildren(object)
+		if err != nil {
+			debugWithState(g, state, "JSON select could not get children: %v (%v)", err, object)
+			return []string{}, nil
+		}
+		var results []string
+		for _, child := range children {
+			innerresults, perr := pvlRunSelectorJSONInner(g, state, child, nextselectors)
+			if perr != nil {
+				return nil, perr
+			}
+			results = append(results, innerresults...)
+		}
+	}
+	return []string{}, libkb.NewProofError(keybase1.ProofStatus_INVALID_PVL,
+		"Selector entry not recognized: %v", selector)
+}
+
+// Take a template, substitute variables, and build the Regexp.
+func pvlInterpretRegex(g ProofContextExt, template string, state PvlScriptState) (*regexp.Regexp, libkb.ProofError) {
+	var perr libkb.ProofError = libkb.NewProofError(keybase1.ProofStatus_INVALID_PVL,
+		"Could not build regex %v", template)
+
+	// Parse out side bars and option letters.
+	if !strings.HasPrefix(template, "/") {
+		return nil, perr
+	}
+	lastSlash := strings.LastIndex(template, "/")
+	if lastSlash == -1 {
+		return nil, perr
+	}
+	opts := template[lastSlash+1:]
+	if !regexp.MustCompile("[imsU]*").MatchString(opts) {
+		return nil, perr
+	}
```

Comment:
Are there any cases where we want to guard against the same letter appearing twice? Or is that just kind of redundant but ultimately harmless?